business_rules/operators.py

```diff
@@ -650,22 +650,54 @@ def is_unique_set(self, other_value):
         return pd.Series(results)
 
     @type_operator(FIELD_DATAFRAME)
-    def is_unique_relationship(self, other_value):
+    def is_not_unique_relationship(self, other_value) -> pd.Series:
+        """
+        Validates one-to-one relationship between two columns (target and comparator) against a dataset.
+        One-to-one means that a pair of columns can be duplicated but its integrity must not be violated:
+        one value of target always corresponds to one value of comparator. Examples:
+
+        Valid dataset:
+        STUDYID  STUDYDESC
+        1        A
+        2        B
+        3        C
+        1        A
+        2        B
+
+        Invalid dataset:
+        STUDYID  STUDYDESC
+        1        A
+        2        A
+        3        C
+        """
         target = self.replace_prefix(other_value.get("target"))
         comparator = other_value.get("comparator")
         if isinstance(comparator, list):
             comparator = self.replace_all_prefixes(comparator)
         else:
             comparator = self.replace_prefix(comparator)
-        grouped_dict = self.value.groupby([comparator])[target].apply(set).to_dict()
-        results = np.where(
-            vectorized_len(vectorized_get_dict_key(grouped_dict, self.value[comparator])) <= 1, True, False
-        )
-        return pd.Series(results)
+        df_without_duplicates: pd.DataFrame = self.value.drop_duplicates()  # remove repeating rows
+        # we need to check if ANY of the columns (target or comparator) is duplicated
+        duplicated_comparator: pd.Series = df_without_duplicates[comparator].duplicated(keep=False)
+        duplicated_target: pd.Series = df_without_duplicates[target].duplicated(keep=False)
+        if duplicated_comparator.any() and not duplicated_target.any():
```

Comment:
Is it possible for more than one of these cases to be triggered for the same dataset?

Reply:
Yes. This case is handled by `elif duplicated_target.any() and duplicated_comparator.any():` condition

Reply:
**UPDATE:** I have refactored the code to make it a bit more elegant and the condition `elif duplicated_target.any() and duplicated_comparator.any():` doesn't exist any more but anyway the case you had mentioned is handled.